### PR TITLE
[tt-train] Don't always fallback to composite RS impl for 2d fabric

### PR DESCRIPTION
**Pull request by philei-tt**
### Summary
TP+DDP llama 8b training has very poor link b/w utilization. The reason is that for some reason when having 2d fabric config all reduce fallbacks to composite reduce scatter implementation that gathers all tensors and then locally reduces. This causes num_devices more bytes to be moved over links. 

- I have remove this fallback, and performance almost doubled! 
- I have also added configurations with bigger batch size that still fit in memory to check our best achievable mfu. Best case => TP=4, DDP=8, B=16 -- 19.7% MFU
- Update outdated ttml distributed docs

Comparison of before/after. 

| Config | Forward (ms) | Backward (ms) | Gradient Sync (ms) | Optimizer (ms) | Other (ms) | Total (ms) | TPS | MFU |
|---|---|---|---|---|---|---|---|---|
| TP=8 DDP=4 B=4 composite | 391.928 | 829.442 | 245.653 | 43.250 | 1.980 | 1512.253 | 5417 | 5.1% |
| TP=4 DDP=8 B=8 composite | 280.023 | 611.037 | 1008.859 | 79.659 | 2.394 | 1981.972 | 8267 | 7.8% |
| TP=8 DDP=4 B=4 fixed | 177.541 | 384.036 | 259.076 | 42.984 | 1.407 | 865.045 | 9470 | 8.9% |
| TP=4 DDP=8 b=8 fixed | 211.037 | 470.429 | 255.934 | 79.037 | 2.214 | 1018.651 | 16085 | 15.2% |
| TP=8 DDP=4 b=12 fixed | 419.119 | 879.917 | 267.048 | 43.152 | 1.698 | 1610.933 | 15255 | 14.3% |
| TP=4 DDP=8 b=16 fixed | 367.923 | 854.775 | 267.944 | 79.518 | 2.255 | 1572.414 | 20840 | 19.7% |
 

Loss seems to be roughly identical (model overfits really good)

image
image

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:philei/fix_tp_plus_DDP_training_perf)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:philei/fix_tp_plus_DDP_training_perf)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:philei/fix_tp_plus_DDP_training_perf)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:philei/fix_tp_plus_DDP_training_perf)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:philei/fix_tp_plus_DDP_training_perf)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:philei/fix_tp_plus_DDP_training_perf)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:philei/fix_tp_plus_DDP_training_perf)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:philei/fix_tp_plus_DDP_training_perf)

### CODE REVIEW

**Review by ntarafdar — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce_async/all_reduce_async.cpp [hunk 1]
```diff
@@ -196,11 +186,7 @@ ttnn::Tensor all_reduce_async(
     const ttnn::Tensor& working_input_tensor =
         interleaved_input_tensor.has_value() ? interleaved_input_tensor.value() : input_tensor;
 
-    const bool composite_for_2d_mesh =
-        tt::tt_fabric::GetFabricConfig() == tt::tt_fabric::FabricConfig::FABRIC_2D &&
-        ttnn::operations::experimental::ccl::detail::is_true_2d_mesh(input_tensor, topology);
-
-    if (composite_all_gather || composite_reduce_scatter || (dim != composite_dim) || composite_for_2d_mesh) {
+    if (composite_all_gather || composite_reduce_scatter || (dim != composite_dim)) {
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce_async/all_reduce_async.cpp:189]: t3k 2d ccl tests should already cover this
